Sources/Core/AWSClientRuntime/EventStream/AWSMessageEncoderStream.swift

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import ClientRuntime
+
+extension AWSEventStream {
+    /// Stream adapter that encodes input into `Data` objects.
+    public class AWSMessageEncoderStream<Element: MessageMarshallable>: MessageEncoderStream {
+        let stream: AsyncThrowingStream<Element, Error>
+        let messageEncoder: MessageEncoder
+        let messageSinger: MessageSigner
+        let requestEncoder: RequestEncoder
+
+        public init(stream: AsyncThrowingStream<Element, Error>,
+                    messageEncoder: MessageEncoder,
+                    requestEncoder: RequestEncoder,
+                    messageSinger: MessageSigner) {
+            self.stream = stream
+            self.messageEncoder = messageEncoder
+            self.messageSinger = messageSinger
+            self.requestEncoder = requestEncoder
+        }
+
+        public struct AsyncIterator: AsyncIteratorProtocol {
+            let stream: AsyncThrowingStream<Element, Error>
+            let messageEncoder: MessageEncoder
+            var messageSinger: MessageSigner
+            let requestEncoder: RequestEncoder
+
+            private var lastMessageSent: Bool = false
+
+            init(stream: AsyncThrowingStream<Element, Error>,
+                 messageEncoder: MessageEncoder,
+                 requestEncoder: RequestEncoder,
+                 messageSinger: MessageSigner) {
+                self.stream = stream
+                self.messageEncoder = messageEncoder
+                self.messageSinger = messageSinger
+                self.requestEncoder = requestEncoder
+            }
+
+            mutating public func next() async throws -> Data? {
+                var iterator = stream.makeAsyncIterator()
```

Comment:
Should we be making this underlying iterator every time `next` is called or does it make more sense to only create this once per instance of this AsyncIterator?

Reply:
not sure we will gain anything here by making per instance. This is basically like doing a for-loop. 

I generally not prefer using an instance index (ie i). If there is strong case which I'm not aware we can consider for sure.